pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,449 @@
+    Step 16: cost = -7.656, shots_used = 41022
+    Step 17: cost = -7.212, shots_used = 48918
+    Step 18: cost = -7.532, shots_used = 56286
+    Step 19: cost = -7.462000000000001, shots_used = 63822
+    Step 20: cost = -7.306000000000001, shots_used = 72534
+    Step 21: cost = -7.234, shots_used = 82014
+    Step 22: cost = -7.314, shots_used = 92838
+    """
+
+    def __init__(self, min_shots, weighted_random_sampling=True, mu=0.99, b=1e-6, stepsize=0.07):
+        self.wrs = weighted_random_sampling
+        self.trainable_args = set()
+
+        # hyperparameters
+        self.min_shots = min_shots
+        self.mu = mu  # running average constant
+        self.b = b  # regularization bias
+        self.lipschitz = None
+
+        # keep track of the total number of shots used
+        self.shots_used = 0
+        # total number of iterations
+        self.k = 0
+        # Number of shots per parameter
+        self.s = None
+        # maximum number of shots allowed per iteration
+        self.max_shots = None
+
+        # Running average of the parameter gradients
+        self.chi = None
+        # Running average of the variance of the parameter gradients
+        self.xi = None
+
+        super().__init__(stepsize=stepsize)
+
+    @staticmethod
+    def weighted_random_sampling(qnodes, coeffs, shots, argnums, *args, **kwargs):
+        """Returns an array containing length ``shots`` single-shot estimates
+        of the Hamiltonian gradient. The shots are distributed randomly over
+        the terms in the Hamiltonian, as per a Multinomial distribution.
+
+        Args:
+            qnodes (Sequence[.tape.QNode]): Sequence of QNodes, each one when evaluated
+                returning the corresponding expectation value of a term in the Hamiltonian.
+            coeffs (Sequence[float]): Sequences of coefficients corresponding to
+                each term in the Hamiltonian.
+            shots (int): The number of shots used to estimate the Hamiltonian expectation
+                value. These shots are distributed over the terms in the Hamiltonian,
+                as per a Multinomial distribution.
+            argnums (Sequence[int]): the QNode argument indices which are trainable
+            *args: Arguments to the QNodes
+            **kwargs: Keyword arguments to the QNodes
+
+        Returns:
+            array[float]: the single-shot gradients of the Hamiltonian expectation value
+        """
+
+        # determine the shot probability per term
+        prob_shots = np.abs(coeffs) / np.sum(np.abs(coeffs))
+
+        # construct the multinomial distribution, and sample
+        # from it to determine how many shots to apply per term
+        si = multinomial(n=shots, p=prob_shots)
+        shots_per_term = si.rvs()[0]
+
+        grads = []
+
+        for h, c, p, s in zip(qnodes, coeffs, prob_shots, shots_per_term):
+
+            # if the number of shots is 0, do nothing
+            if s == 0:
+                continue

```

Comment:
test?

Reply:
I have to admit, I wasn't sure how to test this, since `s` is sampled!

Oh, maybe I can monkeypatch scipy...